modules/renter/hostdb/hosttree/hosttree.go

```diff
@@ -0,0 +1,258 @@
+package hosttree
+
+import (
+	"crypto/rand"
+	"errors"
+
+	"github.com/NebulousLabs/Sia/modules"
+	"github.com/NebulousLabs/Sia/types"
+)
+
+var (
+	// ErrWeightTooHeavy is returned from a Fetch() call if a weight that exceeds
+	// the total weight of the tree is requested.
+	ErrWeightTooHeavy = errors.New("requested a too-heavy weight")
+
+	// ErrNegativeWeight is returned from an Insert() call if an entry with a
+	// negative weight is added to the tree. Entries must always have a positive
+	// weight.
+	ErrNegativeWeight = errors.New("cannot insert using a negative weight")
+
+	// ErrNilEntry is returned if a fetch call results in a nil tree entry. nodes
+	// should always have a non-nil entry, unless they have been Delete()ed.
+	ErrNilEntry = errors.New("node has a nil entry")
+
+	// ErrHostExists is returned if an Insert is called with a public key that
+	// already exists in the tree.
+	ErrHostExists = errors.New("host already exists in the tree")
+
+	// ErrNoSuchHost is returned if Remove is called with a public key that does
+	// not exist in the tree.
+	ErrNoSuchHost = errors.New("no host with specified public key")
+)
+
+type (
+	// HostTree is a data structure that contains a weighted tree of
+	// `HostEntries`, and provides methods for inserting, retreiving, and
+	// modifing entries.
+	HostTree struct {
+		*node
+
+		// hosts is a map of public keys to nodes.
+		hosts map[string]*node
+	}
+
+	// HostEntry is an entry in the host tree.
+	HostEntry struct {
+		modules.HostDBEntry
+
+		FirstSeen   types.BlockHeight
+		Weight      types.Currency
+		Reliability types.Currency
```

Comment:
These could use docstrings. Also, does the caller need to know what the Weight of an entry is?